cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java

```diff
@@ -17,286 +17,210 @@
 import co.cask.cdap.api.spark.SparkSpecification;
-import co.cask.cdap.api.stream.StreamEventDecoder;
-import co.cask.cdap.data.stream.StreamInputFormat;
-import co.cask.cdap.data.stream.StreamUtils;
-import co.cask.cdap.data2.transaction.stream.StreamConfig;
-import co.cask.cdap.internal.app.runtime.batch.dataset.DataSetInputFormat;
-import co.cask.cdap.internal.app.runtime.batch.dataset.DataSetOutputFormat;
-import co.cask.cdap.internal.app.runtime.spark.dataset.SparkDatasetInputFormat;
-import co.cask.cdap.internal.app.runtime.spark.dataset.SparkDatasetOutputFormat;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.logging.LoggingContext;
+import co.cask.cdap.internal.app.program.ProgramTypeMetricTag;
+import co.cask.cdap.internal.app.runtime.spark.metrics.SparkUserMetrics;
+import co.cask.cdap.logging.context.SparkLoggingContext;
 import co.cask.cdap.proto.Id;
-import com.google.gson.Gson;
-import org.apache.hadoop.conf.Configuration;
-import org.apache.hadoop.io.BytesWritable;
-import org.apache.hadoop.io.Text;
-import org.apache.hadoop.mapreduce.InputFormat;
-import org.apache.hadoop.mapreduce.MRJobConfig;
-import org.apache.hadoop.mapreduce.OutputFormat;
-import org.apache.spark.SparkConf;
-import org.apache.spark.api.java.JavaPairRDD;
-import org.apache.spark.rdd.NewHadoopRDD;
-import org.apache.twill.filesystem.Location;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import java.io.IOException;
-import java.net.URL;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import org.apache.twill.api.RunId;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+import java.io.Closeable;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 /**
- * An abstract class which implements {@link SparkContext} and provide a concrete implementation for the common
- * functionality between {@link JavaSparkContext} and {@link ScalaSparkContext}
+ * An abstract implementation of {@link SparkContext} for common functionality that spread across
+ * execution context for {@link Spark} and {@link SparkProgram}.
  */
-abstract class AbstractSparkContext implements SparkContext {
+public abstract class AbstractSparkContext implements SparkContext, Closeable {
 
-  private static final Logger LOG = LoggerFactory.getLogger(AbstractSparkContext.class);
-  private static final Gson GSON = new Gson();
-  private static final Pattern SPACES = Pattern.compile("\\s+");
-  private static final String[] NO_ARGS = {};
-  private static final String SPARK_METRICS_CONF_KEY = "spark.metrics.conf";
+  // A Regex pattern that split with whitespace that supports "" quotation.
+  private static final Pattern QUOTED_STRING_PATTERN = Pattern.compile("([^\"]\\S*|\".+?\")\\s*");
```

Comment:
Not sure what that method is for? It takes the value for a given key, and splits it? So I could have a runtime argument 
myKey=abc "x y z" jkl, and this will return ["abc", "x y z", "jkl"] ? What is a use case for such an API?